error-codes/README.md

```diff
@@ -0,0 +1,105 @@
+QUIC provides the ability to send an error on closing the read end of the
+stream, resetting the write end of the stream and on closing the connection. 
+
+For stream resets, the error code MUST be sent on `RESET_STREAM` and `STOP_SENDING` frames using the `Application Protocol Error Code` field as per
+the frame definitions in the
+[RFC](https://www.rfc-editor.org/rfc/rfc9000.html#name-reset_stream-frames).
+
+For Connection Close, the error code MUST be sent on `CONNECTION_CLOSE` frame
+using the Error Code field as defined in the
+[RFC](https://www.rfc-editor.org/rfc/rfc9000.html#section-19.19-6.2.1).
+
+### Yamux
+Yamux streams are reset unilaterally. Receiving a stream frame with `RST` flag set resets both the read and write end of the stream. So, there's no way to separately signal error code on closing the read end of the stream, or resetting the write end of the stream. 
+
+For Connection Close, the 32-bit Length field is interpreted as the error code.
+
+For Stream Resets, the error code is sent in the `Window Update` frame, with the
+32-bit Length field interpreted as the error code. See [yamux spec
+extension](https://github.com/libp2p/specs/pull/622).
+
+Connection Close error code delivery to the other end depends on the OS TCP implementation and the TCP options used for the socket. In particular, when `SO_LINGER` TCP option is set to 0 and the implementation closes the connection immediately after writing the error code containing frame, the error code may not be delivered.
+
+### WebRTC
+There is no way to provide any information on closing a peer connection in
+WebRTC. Providing error codes on Connection Close will be taken up in the
+future. 
+
+For Stream Resets, the error code can be sent in the `errorCode` field of the
+WebRTC message with `flag` set to `RESET_STREAM`.
+
+### WebTransport
+Error codes for WebTransport will be introduced when browsers upgrade to draft-9
+of the spec. The current WebTransport spec implemented by Chrome and Firefox is
+[draft-2 of WebTransport over
+HTTP/3](https://www.ietf.org/archive/id/draft-ietf-webtrans-http3-02.html#section-4.3-2).
+This version allows for only a 1-byte error code. 1 byte is too restrictive and
+as the latest WebTransport draft,
+[draft-9](https://www.ietf.org/archive/id/draft-ietf-webtrans-http3-02.html#section-4.3-2)
+allows for a 4-byte error code to be sent on stream resets, we will introduce
+error codes over WebTransport later.
+
+### HTTP
+Protocols that work over http MUST use the response header `Libp2p-Error-Code` to send the error code. The grammar for the field is similar to `Content-Length`
+```
+Libp2p-Error-Code: 1*DIGIT
+```
+
+### Multistream Select
+Multistream-Select is used to negotiate Security protocol for TCP connections before a stream muxer has been selected. There's only one error code defined for such cases, `PROTOCOL_NEGOTIATION_FAILED`. To encode this error, send the string `101` prefixed with the length and close the TCP connection.  
```

Comment:
This is not too useful. The most common case here is a server rejecting a connection because the handling it exceeds some resource limit. In such cases Writing on the connection and closing will cause the write to be dropped because almost certainly there will be unread data in the read buffer.